## Week 5 Homework 

In this homework we'll put what we learned about Spark in practice.

For this homework we will be using the FHVHV 2021-06 data found here. [FHVHV Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz )


### Question 1: 

**Install Spark and PySpark** 

- Install Spark
- Run PySpark
- Create a local spark session
- Execute spark.version.


In [2]:
import warnings
warnings.filterwarnings("ignore")

In [3]:
import pyspark

In [4]:
from pyspark.sql import SparkSession

In [5]:
spark = SparkSession.builder \
    .master("local[*]") \
    .appName('homework') \
    .getOrCreate()

23/03/08 18:36:53 WARN Utils: Your hostname, MacBook-Air-de-Lucas.local resolves to a loopback address: 127.0.0.1; using 192.168.1.108 instead (on interface en0)
23/03/08 18:36:53 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
23/03/08 18:36:54 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
23/03/08 18:36:56 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
23/03/08 18:36:56 WARN Utils: Service 'SparkUI' could not bind on port 4041. Attempting port 4042.
23/03/08 18:36:56 WARN Utils: Service 'SparkUI' could not bind on port 4042. Attempting port 4043.


In [6]:
spark.version

'3.2.1'

What's the output?
- ***3.3.2***
- 2.1.4
- 1.2.3
- 5.4
</br></br>

### Question 2: 

**HVFHW June 2021**

Read it with Spark using the same schema as we did in the lessons.</br> 
We will use this dataset for all the remaining questions.</br>
Repartition it to 12 partitions and save it to parquet.</br>
What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.</br>

In [12]:
!wget 'https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz'

--2023-03-07 18:09:34--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz
Resolving github.com (github.com)... 20.201.28.151
Connecting to github.com (github.com)|20.201.28.151|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/4564ad9e-a6da-4923-ad6f-35ff02446a51?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230307%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230307T210935Z&X-Amz-Expires=300&X-Amz-Signature=939a4a6e2283671f818ed9baa2177cde652a9145b8e4b8f81fcf54fc8716076e&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhvhv_tripdata_2021-06.csv.gz&response-content-type=application%2Foctet-stream [following]
--2023-03-07 18:09:34--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/4564ad

In [15]:
!wc -l fhvhv_tripdata_2021-06.csv.gz

  651315 fhvhv_tripdata_2021-06.csv.gz


In [7]:
import pandas as pd

In [16]:
df_pandas = pd.read_csv('fhvhv_tripdata_2021-06.csv.gz', nrows=100)

In [17]:
df_pandas.head()

dispatching_base_num      pickup_datetime     dropoff_datetime  \
0               B02764  2021-06-01 00:02:41  2021-06-01 00:07:46   
1               B02764  2021-06-01 00:16:16  2021-06-01 00:21:14   
2               B02764  2021-06-01 00:27:01  2021-06-01 00:42:11   
3               B02764  2021-06-01 00:46:08  2021-06-01 00:53:45   
4               B02510  2021-06-01 00:45:42  2021-06-01 01:03:33   

   PULocationID  DOLocationID SR_Flag Affiliated_base_number  
0           174            18       N                 B02764  
1            32           254       N                 B02764  
2           240           127       N                 B02764  
3           127           235       N                 B02764  
4           144           146       N                    NaN

In [19]:
df_pandas = pd.read_csv('fhvhv_tripdata_2021-06.csv.gz',
                        parse_dates=["pickup_datetime","dropoff_datetime"], 
                        nrows=1000)

In [20]:
df_pandas.dtypes

dispatching_base_num              object
pickup_datetime           datetime64[ns]
dropoff_datetime          datetime64[ns]
PULocationID                       int64
DOLocationID                       int64
SR_Flag                           object
Affiliated_base_number            object
dtype: object

In [23]:
df_pandas.isnull().sum()

dispatching_base_num        0
pickup_datetime             0
dropoff_datetime            0
PULocationID                0
DOLocationID                0
SR_Flag                     0
Affiliated_base_number    265
dtype: int64

In [25]:
df_pandas = df_pandas[df_pandas['Affiliated_base_number'].notna()]

In [26]:
spark.createDataFrame(df_pandas)

DataFrame[dispatching_base_num: string, pickup_datetime: timestamp, dropoff_datetime: timestamp, PULocationID: bigint, DOLocationID: bigint, SR_Flag: string, Affiliated_base_number: string]

In [27]:
spark.createDataFrame(df_pandas).schema

StructType(List(StructField(dispatching_base_num,StringType,true),StructField(pickup_datetime,TimestampType,true),StructField(dropoff_datetime,TimestampType,true),StructField(PULocationID,LongType,true),StructField(DOLocationID,LongType,true),StructField(SR_Flag,StringType,true),StructField(Affiliated_base_number,StringType,true)))

In [8]:
from pyspark.sql import types

In [29]:
schema = types.StructType([
    types.StructField('dispatching_base_num', types.StringType(), True), 
    types.StructField('pickup_datetime', types.TimestampType(), True), 
    types.StructField('dropoff_datetime', types.TimestampType(), True), 
    types.StructField('PULocationID', types.IntegerType(), True), 
    types.StructField('DOLocationID', types.IntegerType(), True), 
    types.StructField('SR_Flag', types.StringType(), True), 
    types.StructField('Affiliated_base_number', types.StringType(), True),
])

In [32]:
df = spark.read \
    .option("header", "true") \
    .schema(schema) \
    .csv('fhvhv_tripdata_2021-06.csv.gz')

In [33]:
df.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B02764|2021-06-01 00:02:41|2021-06-01 00:07:46|         174|          18|      N|                B02764|
|              B02764|2021-06-01 00:16:16|2021-06-01 00:21:14|          32|         254|      N|                B02764|
|              B02764|2021-06-01 00:27:01|2021-06-01 00:42:11|         240|         127|      N|                B02764|
|              B02764|2021-06-01 00:46:08|2021-06-01 00:53:45|         127|         235|      N|                B02764|
|              B02510|2021-06-01 00:45:42|2021-06-01 01:03:33|         144|         146|      N|                  null|
|              B02510|2021-06-01 00:18:1

In [34]:
df.head(5)

[Row(dispatching_base_num='B02764', pickup_datetime=datetime.datetime(2021, 6, 1, 0, 2, 41), dropoff_datetime=datetime.datetime(2021, 6, 1, 0, 7, 46), PULocationID=174, DOLocationID=18, SR_Flag='N', Affiliated_base_number='B02764'),
 Row(dispatching_base_num='B02764', pickup_datetime=datetime.datetime(2021, 6, 1, 0, 16, 16), dropoff_datetime=datetime.datetime(2021, 6, 1, 0, 21, 14), PULocationID=32, DOLocationID=254, SR_Flag='N', Affiliated_base_number='B02764'),
 Row(dispatching_base_num='B02764', pickup_datetime=datetime.datetime(2021, 6, 1, 0, 27, 1), dropoff_datetime=datetime.datetime(2021, 6, 1, 0, 42, 11), PULocationID=240, DOLocationID=127, SR_Flag='N', Affiliated_base_number='B02764'),
 Row(dispatching_base_num='B02764', pickup_datetime=datetime.datetime(2021, 6, 1, 0, 46, 8), dropoff_datetime=datetime.datetime(2021, 6, 1, 0, 53, 45), PULocationID=127, DOLocationID=235, SR_Flag='N', Affiliated_base_number='B02764'),
 Row(dispatching_base_num='B02510', pickup_datetime=datetime.d

In [35]:
df = df.repartition(12)

In [36]:
df.write.parquet('fhvhv/2021/06/')

In [37]:
df = spark.read.parquet('fhvhv/2021/06/')

In [39]:
!ls -lh fhvhv/2021/06

total 585016
-rw-r--r--  1 lukas17e  staff     0B Mar  7 18:49 _SUCCESS
-rw-r--r--  1 lukas17e  staff    24M Mar  7 18:49 part-00000-58af1777-d4c0-4ad8-ac53-f504e644f802-c000.snappy.parquet
-rw-r--r--  1 lukas17e  staff    24M Mar  7 18:49 part-00001-58af1777-d4c0-4ad8-ac53-f504e644f802-c000.snappy.parquet
-rw-r--r--  1 lukas17e  staff    24M Mar  7 18:49 part-00002-58af1777-d4c0-4ad8-ac53-f504e644f802-c000.snappy.parquet
-rw-r--r--  1 lukas17e  staff    24M Mar  7 18:49 part-00003-58af1777-d4c0-4ad8-ac53-f504e644f802-c000.snappy.parquet
-rw-r--r--  1 lukas17e  staff    24M Mar  7 18:49 part-00004-58af1777-d4c0-4ad8-ac53-f504e644f802-c000.snappy.parquet
-rw-r--r--  1 lukas17e  staff    24M Mar  7 18:49 part-00005-58af1777-d4c0-4ad8-ac53-f504e644f802-c000.snappy.parquet
-rw-r--r--  1 lukas17e  staff    24M Mar  7 18:49 part-00006-58af1777-d4c0-4ad8-ac53-f504e644f802-c000.snappy.parquet
-rw-r--r--  1 lukas17e  staff    24M Mar  7 18:49 part-00007-58af1777-d4c0-4ad8-ac53-f504e644f802-c000

- 2MB
- ***24MB***
- 100MB
- 250MB
</br></br>

### Question 3: 

**Count records**  

How many taxi trips were there on June 15?</br></br>
Consider only trips that started on June 15.</br>

In [9]:
df = spark.read.parquet('fhvhv/2021/06/')

In [10]:
df.registerTempTable('trips_data')

In [20]:
spark.sql("""
SELECT 
    date_trunc('DD', pickup_datetime) AS day,
    COUNT(1) AS trips_records
FROM
    trips_data
WHERE
    pickup_datetime >= '2021-06-15'
AND
    pickup_datetime < '2021-06-16'
GROUP BY
    1
""").show()

+-------------------+-------------+
|                day|trips_records|
+-------------------+-------------+
|2021-06-15 00:00:00|       452470|
+-------------------+-------------+



- 308,164
- 12,856
- ***452,470***
- 50,982
</br></br>

### Question 4: 

**Longest trip for each day**  

Now calculate the duration for each trip.</br>
How long was the longest trip in Hours?</br>

In [21]:
from pyspark.sql.functions import unix_timestamp, col

In [27]:
spark.sql("""
SELECT dispatching_base_num,
        pickup_datetime,
        dropoff_datetime, 
    (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime))/3600 AS duration
    FROM trips_data
    ORDER BY duration DESC
    LIMIT 1
""").show()

+--------------------+-------------------+-------------------+----------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|        duration|
+--------------------+-------------------+-------------------+----------------+
|              B02872|2021-06-25 13:55:41|2021-06-28 08:48:25|66.8788888888889|
+--------------------+-------------------+-------------------+----------------+



- ***66.87 Hours***
- 243.44 Hours
- 7.68 Hours
- 3.32 Hours
</br></br>

### Question 5: 

**User Interface**

 Spark’s User Interface which shows application's dashboard runs on which local port?</br>

In [ ]:
spark-shell
23/03/06 16:31:07 WARN Utils: Your hostname, MacBook-Air-de-Lucas.local resolves to a loopback address: 127.0.0.1; using 192.168.1.108 instead (on interface en0)
23/03/06 16:31:07 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
23/03/06 16:31:15 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Spark context Web UI available at http://192.168.1.108:4040
Spark context available as 'sc' (master = local[*], app id = local-1678131077184).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.3.2
      /_/
         
Using Scala version 2.12.15 (OpenJDK 64-Bit Server VM, Java 11.0.18)
Type in expressions to have them evaluated.
Type :help for more information.

- 80
- 443
- ***4040***
- 8080
</br></br>

### Question 6: 

**Most frequent pickup location zone**

Load the zone lookup data into a temp view in Spark</br>
[Zone Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv)</br>

Using the zone lookup data and the fhvhv June 2021 data, what is the name of the most frequent pickup location zone?</br>

In [29]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv

--2023-03-08 19:35:07--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv
Resolving github.com (github.com)... 20.201.28.151
Connecting to github.com (github.com)|20.201.28.151|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6ea97ed0e6a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230308%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230308T223507Z&X-Amz-Expires=300&X-Amz-Signature=0472a6413d50f335f1be30605a528ba4af1b789d955687170e06468b6a56204c&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dtaxi_zone_lookup.csv&response-content-type=application%2Foctet-stream [following]
--2023-03-08 19:35:07--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a

In [30]:
zones_df = spark.read \
    .option("header", "true") \
    .csv('taxi_zone_lookup.csv')

In [31]:
zones_df.write.parquet('zones', mode='overwrite')

In [50]:
df_zones = spark.read.parquet('zones')

In [51]:
df_zones.show()

+----------+-------------+--------------------+------------+
|LocationID|      Borough|                Zone|service_zone|
+----------+-------------+--------------------+------------+
|         1|          EWR|      Newark Airport|         EWR|
|         2|       Queens|         Jamaica Bay|   Boro Zone|
|         3|        Bronx|Allerton/Pelham G...|   Boro Zone|
|         4|    Manhattan|       Alphabet City| Yellow Zone|
|         5|Staten Island|       Arden Heights|   Boro Zone|
|         6|Staten Island|Arrochar/Fort Wad...|   Boro Zone|
|         7|       Queens|             Astoria|   Boro Zone|
|         8|       Queens|        Astoria Park|   Boro Zone|
|         9|       Queens|          Auburndale|   Boro Zone|
|        10|       Queens|        Baisley Park|   Boro Zone|
|        11|     Brooklyn|          Bath Beach|   Boro Zone|
|        12|    Manhattan|        Battery Park| Yellow Zone|
|        13|    Manhattan|   Battery Park City| Yellow Zone|
|        14|     Brookly

In [71]:
df_result = df.join(df_zones, df.PULocationID == df_zones.LocationID)

In [72]:
df_result.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+----------+---------+--------------------+------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|LocationID|  Borough|                Zone|service_zone|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+----------+---------+--------------------+------------+
|              B02510|2021-06-04 08:13:36|2021-06-04 08:47:58|          91|          65|      N|                  null|        91| Brooklyn|           Flatlands|   Boro Zone|
|              B02875|2021-06-25 01:52:01|2021-06-25 02:14:51|         132|         255|      N|                B02875|       132|   Queens|         JFK Airport|    Airports|
|              B02875|2021-06-25 03:34:13|2021-06-25 03:42:22|          75|         263|      N|                B02875|      

In [73]:
df_result \
    .drop('LocationID') \
    .write.parquet('tmp/revenue-zones', mode='overwrite')

In [74]:
df_result = spark.read.parquet('tmp/revenue-zones')

In [75]:
df_result.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+-------------+--------------------+------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|      Borough|                Zone|service_zone|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+-------------+--------------------+------------+
|              B02875|2021-06-17 01:00:11|2021-06-17 01:14:43|          50|          42|      N|                B02875|    Manhattan|        Clinton West| Yellow Zone|
|              B02888|2021-06-10 20:17:25|2021-06-10 20:54:59|         137|         227|      N|                B02888|    Manhattan|            Kips Bay| Yellow Zone|
|              B02764|2021-06-22 14:47:18|2021-06-22 15:02:13|         186|          50|      N|                B02764|    Manhattan|Penn Station/Madi...| Yello

In [76]:
df_result.registerTempTable('join_data')

In [78]:
spark.sql("""
SELECT
    Zone,
    COUNT(1) as total_trips
FROM 
    join_data
GROUP BY 1
ORDER BY 2 DESC
LIMIT 1
""").show()

+-------------------+-----------+
|               Zone|total_trips|
+-------------------+-----------+
|Crown Heights North|     231279|
+-------------------+-----------+



- East Chelsea
- Astoria
- Union Sq
- ***Crown Heights North***
</br></br>

## Submitting the solutions

* Form for submitting: https://forms.gle/EcSvDs6vp64gcGuD8
* You can submit your homework multiple times. In this case, only the last submission will be used. 

Deadline: 06 March (Monday), 22:00 CET


## Solution

